Update help

azangru opened the pull request:
## Description
This PR contains the following changes:
1) Updated the contextual help popup code to support the backend changes (see [ENSWBSITES-930](https://www.ebi.ac.uk/panda/jira/browse/ENSWBSITES-930))
  - related articles are now different for different articles — huge improvement for code readability
  - video articles are now articles in their own right
2) Added the `back` and `forward` history buttons for the contextual help popup (see [ENSWBSITES-1052](https://www.ebi.ac.uk/panda/jira/browse/ENSWBSITES-1052)). They became necessary because now each article has its own set of related articles (or may have none), so the user needs a way to return to the article he was viewing before he clicked on a related article.
3) Added the About Ensembl page (can be accessed by clicking on the "About Ensembl" image in the top right corner of the screen, see [ENSWBSITES-931](https://www.ebi.ac.uk/panda/jira/browse/ENSWBSITES-931)).

## Deployment URL
http://update-help.review.ensembl.org/

## Views affected
- Contextual help popup
- About Ensembl page

## Left to be done in a subsequent PR
- VideoArticle component to be extracted from the contextual help popup
- TextArticle component (currently used on the About Ensembl page) to be re-used in the contextual help popup
- Help app to be added, which will re-use both the VideoArticle and the TextArticle components

## To code reviewers
If you try to run the code in this branch locally, notice that you will need to update the proxy in the webpack config to route all requests for `/api/docs` to http://update-help.review.ensembl.org/api/docs/

azangru commented on src/ensembl/webpack/environments/webpack.config.dev.js:
```diff
@@ -29,14 +29,37 @@ const devServerConfig = {
+      target: 'http://localhost:3000',
+      pathRewrite: { '^/api/docs': '/api' },
+      changeOrigin: true,
+      secure: false
```
**Note to self:** remember to remove this proxy rule (and update the main rule not to exclude the docs) before the merge

msmanoj commented:
Wouldn't it be nice to add a similar styling to the menu links to show the active page like it does on the side menu?
![Kapture 2021-03-25 at 00 20 38](https://user-images.githubusercontent.com/3085815/112400349-25492880-8d00-11eb-992d-60885a714f6c.gif)

msmanoj commented on src/ensembl/src/shared/types/help-and-docs/menu.ts:
```diff
@@ -0,0 +1,35 @@
+type MenuCollectionItem = {
+  name: string;
+  type: 'collection';
+  url?: string; // if a meenu directory has an index page associated with it
```
`meenu` -> `menu`

azangru replied:
Damn this butterfly Macbook keyboard! 😠  It's keeps doing this to me, spontaneously duplicating the letter e and a couple of others whenever it feels like it.

msmanoj replied:
Haha! Looks like I am not alone with this :D

msmanoj commented on src/ensembl/src/shared/types/help-and-docs/menu.ts:
```diff
@@ -0,0 +1,35 @@
+
+export type Menu = {
+  name: string;
+  data: MenuItem[];
```
Wouldn't `items` be a better name instead of `data`?

azangru replied:
Renamed `data` to `items`.

azangru commented:
> Wouldn't it be nice to add a similar styling to the menu links to show the active page like it does on the side menu?

Yes. I was sure that I had checked with the [XD](https://xd.adobe.com/view/d9d03d0b-700c-4e77-a937-2a88de5336bb-8c83/screen/4bc0093f-3d5b-4b3a-be49-b373b560a5fa?fullscreen) and that is did not style the top menu, but I was wrong.

msmanoj commented on src/ensembl/src/shared/components/help-article/HelpArticle.scss:
```diff
@@ -0,0 +1,41 @@
+
+  // Notice that the external-link svg has the same shape as in the static/img/shared folder
+  // But the version copied here has the fill color set inline, because it is being accessed as an external image via url
+  a[href^=http] {
```
Won't these catch internal links as well?

azangru replied:
Depends. If an internal link is just a link to another help article, it will likely be written in the markdown source file like this: `[Another article]('another-article.md')`, which will get transformed into something like `/help/articles/another-article`, and thus won't have a protocol in it.

Links to other sections of the site can also be written without protocol and host (e.g. `/browser`, `/entity-viewer`). This will leave the full urls, with protocol and host written out, only for truly external links.

msmanoj commented on src/ensembl/src/shared/components/help-article/VideoArticle.tsx:
```diff
@@ -0,0 +1,15 @@
+/**
```
Do we need to keep this file?

azangru replied:
Eventually yes, but in this PR probably not. I'll remove it from this PR and add it back later.

azangru replied:
Removed.

msmanoj commented on src/ensembl/src/shared/components/help-popup/types.ts:
```diff
@@ -14,31 +14,37 @@
+};
+
+export type TextArticle = Article & {
+  type: 'article';
```
Wouldn't it be better to call this something like `text-article` instead? 
I feel it's slightly confusing to have Article of type `article`.

azangru replied:
Umm. This is tightly linked to what the help&docs api is returning. Compare the responses:

- [for a text article](http://update-help.review.ensembl.org/api/docs/article?slug=species-selector)
- [for a video article](http://update-help.review.ensembl.org/api/docs/article?slug=ensembl-help-getting-started-videos-find-and-add-species)

The api is ours, and we can of course change what it responds with, but the `type: "video"` in the response just feels so much nicer than something like `type: "video_article"`...

msmanoj replied:
As discussed during the code review, we have decided to leave the current implementation as it is.

msmanoj commented on src/ensembl/src/shared/components/help-popup/useHelpPopupHistory.ts:
```diff
@@ -0,0 +1,15 @@
+/**
```
I guess this file can also be created when it is actually needed.

azangru replied:
✅

msmanoj commented on src/ensembl/src/shared/types/help-and-docs/article.ts:
```diff
@@ -0,0 +1,42 @@
+};
+
+export type TextArticle = Article & {
+  type: 'article';
```
Similar to one of the previous comments regarding changing this to `text-article`

azangru commented:
@msmanoj I've updated the top menu, renamed `data` field of the menu to `items`, deleted the unused files and fixed the typo in the comment. I'm reluctant to change the `article` and `video` type labels for articles to something like `text-article` and `video-article`. We might change them to `text` and `video` if we like this better than `article` and `video`; but since this will also require certain changes on the backend, I propose to leave it for a separate PR.

imransl commented on src/ensembl/src/content/app/about/components/about-menu/AboutMenu.tsx:
```diff
@@ -0,0 +1,91 @@
+export const TopMenu = (props: Props) => {
+  const { menu } = props;
+
+  const menuItems = menu.items;
```
Do we need to create this variable? In line 55 `menu.items` is used without assigning it to to a new variable/constant.

azangru replied:
🤷‍♂️  ✅

imransl commented on src/ensembl/src/content/app/about/components/about-menu/AboutMenu.tsx:
```diff
@@ -0,0 +1,91 @@
+};
+
+const hasItemWithUrl = (menuItem: MenuItem, url: string) => {
+  if (menuItem.url && menuItem.url === url) {
```
Isn't checking if `menuItem.url` is falsey redundant since we are checking if it matches `url` in the `if` condition?

azangru replied:
Yep, this can be removed. Done.

azangru commented on src/ensembl/src/content/app/about/components/about-menu/AboutMenu.tsx:
```diff
@@ -0,0 +1,91 @@
+  } else if (menuItem.type === 'collection' && menuItem.items.length) {
+    if (menuItem.items.some((item) => hasItemWithUrl(item, url))) {
+      return true;
+    }
```
👍  ✅

azangru commented on src/ensembl/src/shared/components/help-article/TextArticle.tsx:
```diff
@@ -0,0 +1,72 @@
+import { TextArticle as TextArticleType } from 'src/shared/types/help-and-docs/article';
+import { useDispatch } from 'react-redux';
+import { push } from 'connected-react-router';
+
```
✅

imransl commented:
> @msmanoj I've updated the top menu, renamed `data` field of the menu to `items`, deleted the unused files and fixed the typo in the comment. I'm reluctant to change the `article` and `video` type labels for articles to something like `text-article` and `video-article`. We might change them to `text` and `video` if we like this better than `article` and `video`; but since this will also require certain changes on the backend, I propose to leave it for a separate PR.

I think we should do this in a separate PR as we rely on backend changes.
